libraries/chain/finality_core.cpp

```diff
@@ -0,0 +1,362 @@
+      // and therefore it is impossible to satisfy the precondition. So going forward, it is safe to assume refs.empty() == false.
+
+      const size_t ref_index = block_num - last_final_block_num();
+
+      // By the precondition, 0 <= ref_index < (current_block_num() - last_final_block_num()).
+      // Then, by invariant 8, 0 <= ref_index < refs.size().
+
+      assert(ref_index < refs.size()); // Satisfied by justification above.
+
+      return refs[ref_index];
+      // By invariants 4 and 6, tail[ref_index].block_num() == block_num, which satisfies the post-condition.
+   }
+
+   /**
+    *  @pre links.front().source_block_num <= block_num <= current_block_num()
+    *
+    *  @post returned qc_link has source_block_num == block_num
+    */
+   const qc_link& finality_core::get_qc_link_from(block_num_type block_num) const
+   {
+      assert(!links.empty()); // Satisfied by invariant 1.
+
+      assert(links.front().source_block_num <= block_num); // Satisfied by precondition.
+      assert(block_num <= current_block_num()); // Satisfied by precondition.
+
+      const size_t link_index = block_num - links.front().source_block_num;
+
+      // By the precondition, 0 <= link_index <= (current_block_num() - links.front().source_block_num).
+      // Then, by invariant 9, 0 <= link_index <= links.size() - 1
+
+      assert(link_index < links.size()); // Satisfied by justification above.
+
+      return links[link_index];
+      // By invariants 7, links[link_index].source_block_num == block_num, which satisfies the post-condition.
+   }
+
+   /**
+    *  @pre current_block.block_num() == this->current_block_num()
+    *  @pre If this->refs.empty() == false, then current_block is the block after the one referenced by this->refs.back()
+    *  @pre this->latest_qc_claim().block_num <= most_recent_ancestor_with_qc.block_num <= this->current_block_num()
+    *  @pre this->latest_qc_claim() <= most_recent_ancestor_with_qc ( (this->latest_qc_claim().block_num == most_recent_ancestor_with_qc.block_num) && most_recent_ancestor_with_qc.is_strong_qc ). When block_num is the same, most_recent_ancestor_with_qc must be stronger than latest_qc_claim()
+    *
+    *  @post returned core has current_block_num() == this->current_block_num() + 1
+    *  @post returned core has latest_qc_claim() == most_recent_ancestor_with_qc
+    *  @post returned core has final_on_strong_qc_block_num >= this->final_on_strong_qc_block_num
+    *  @post returned core has last_final_block_num() >= this->last_final_block_num()
+    */
+   finality_core finality_core::next(const block_ref& current_block, const qc_claim& most_recent_ancestor_with_qc) const
+   {
+      assert(current_block.block_num() == current_block_num()); // Satisfied by precondition 1.
+      
+      assert(refs.empty() || (refs.back().timestamp <= current_block.timestamp)); // Satisfied by precondition 2.
```

Comment:
In Areg's original algorithm, the timestamp comparison was `<`. I run into cases where timestamp were the same. I am not sure if the change of '<=' was correct.